crypto/bls/bls.go

```diff
@@ -18,14 +22,30 @@ func RandPrivateKey() *bls.SecretKey {
 	return &sec
 }
 
+var (
+	errEmptyByte = errors.New("BytesToBLSPublicKey: byte is empty")
+	errCastError = errors.New("BytesToBLSPublicKey: cast type error")
+)
+
 // BytesToBLSPublicKey converts bytes into bls.PublicKey pointer.
 func BytesToBLSPublicKey(bytes []byte) (*bls.PublicKey, error) {
 	if len(bytes) == 0 {
-		return nil, fmt.Errorf("[BytesToBLSPublicKey] bytes is empty")
+		return nil, errEmptyByte
+	}
+	kkey := string(bytes)
+	key, err, _ := BLSPubKeyCache.Do(kkey, func() (interface{}, error) {
```

Comment:
If now we are just caching a simple data results. Better use lru cache with size such as 1000. It's much simpler in logic and probably also in performance. SingleFlight is more for complicated function call caching.

Reply:
make sense.